bootloader/src/old_machine_bootloader_linux.c

```diff
@@ -0,0 +1,290 @@
+                    result_string[j] = replacement[j - i];
+                }
+                substringReplaced = true;
+            }
+        }
+    }
+    result_string[result_size] = '\0';
+    if(substringReplaced){
+        return result_string;
+    } else {
+        return str;
+    }
+}
+
+char* concatenate(size_t size, char *array[size], const char *joint){
+    size_t jlen, lens[size];
+    size_t i, total_size = (size-1) * (jlen=strlen(joint)) + 1;
+    char *result, *p;
+    for(i=0;i<size;++i){
+        total_size += (lens[i]=strlen(array[i]));
+    }
+    p = result = malloc(total_size);
+    for(i=0;i<size;++i){
+        printf("%s\n", array[i]);
+        memcpy(p, array[i], lens[i]);
+        p += lens[i];
+        if(i<size-1){
+            memcpy(p, joint, jlen);
+            p += jlen;
+        }
+    }
+    *p = '\0';
+    return result;
+}
+
+struct response sendRequest(char* server, char* tunnel, bool tunnelUsed, char* data){
+    CURL *curl;
+    CURLcode res;
+    struct response s;
+    curl = curl_easy_init();
+    init_response(&s);
+    if(curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, server);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writefunc);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &s);
+        if(tunnelUsed){
+            curl_easy_setopt(curl, CURLOPT_PROXY, tunnel);
+        }
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+        /* Perform the request, res will get the return code */
+        res = curl_easy_perform(curl);
+        /* Check for errors */
+        if(res != CURLE_OK){
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+            s.ptr = "FAILED";
+            s.len = strlen(s.ptr);
+            return s;
+        }
+
+        /* always cleanup */
+        curl_easy_cleanup(curl);
+    }
+    curl_global_cleanup();
+    return s;
+}
+
+char** getIpAddresses(int maxSize, int *addrCount){
+    char** IPs = malloc(maxSize * sizeof(char*));
+    int i = 0;
+
+    struct ifaddrs * ifAddrStruct=NULL;
+    struct ifaddrs * ifa=NULL;
+    void * tmpAddrPtr=NULL;
+    getifaddrs(&ifAddrStruct);
+
+    for (ifa = ifAddrStruct; ifa != NULL; ifa = ifa->ifa_next) {
+        if (!ifa->ifa_addr || i >= maxSize) {
+            continue;
+        }
+        if (ifa->ifa_addr->sa_family == AF_INET) { // check it is IP4
+            // is a valid IP4 Address
+            tmpAddrPtr=&((struct sockaddr_in *)ifa->ifa_addr)->sin_addr;
+            IPs[i] = malloc(INET_ADDRSTRLEN);
+            inet_ntop(AF_INET, tmpAddrPtr, IPs[i], INET_ADDRSTRLEN);
+            i = i+1;
+        }
+    }
+    if (ifAddrStruct!=NULL) freeifaddrs(ifAddrStruct);
+    *addrCount = i;
+    return IPs;
+}
+
+int ping_island(int argc, char * argv[])
+{
+    // Get system info
+    struct utsname systemInfo;
+
+    // Get glibc version
+    char* glibcVersion = executeCommand("ldd --version");
```

Comment:
Why are we not using GNU_GET_LIBC_VERSION? What if ldd is not installed, for example, in a stripped down server?

Reply:
If gnu_get_libc_version doesn't exist we're screwed anyway

Reply:
If we compile bootloader with a static libc it will always return the libc it was compiled with instead of the one on the os.